lib/projectassignment/project_assignment.go

```diff
@@ -22,22 +34,38 @@ func EnsureProjectAssignmentAuthorized(ctx context.Context, authorizer engine.V2
 	}
 
 	var unauthorizedProjects []string
-	for _, proj := range projectIDs {
+	for _, proj := range projectsToAuthz {
 		if !authorized[proj] {
 			unauthorizedProjects = append(unauthorizedProjects, proj)
 		}
 	}
 
 	if len(unauthorizedProjects) != 0 {
-		return NewProjectsUnauthorizedForAssignmentError(unauthorizedProjects)
+		return NewProjectsUnauthorizedForAssignmentError(unauthorizedProjects, subjects)
 	}
 	return nil
 }
 
-// CalculateProjectDiff returns the symmetric difference of oldProjects and newProjects,
+// calculateProjectsToAuthorize returns the symmetric difference of oldProjects and newProjects,
 // meaning that any project that is not in the intersection of oldProjects and newProjects
-// will be returned.
-func CalculateProjectDiff(oldProjects []string, newProjects []string) []string {
+// will be returned. It also accounts for the internal understanding that if either the field for
+// new or the field for old projects is empty, we need to include '(unassigned)' as part of
+// the list of projects requiring authorization.
+func calculateProjectsToAuthorize(oldProjects, newProjects []string, isUpdateRequest bool) []string {
+	projectDiff := []string{}
+	previouslyUnassigned := len(oldProjects) == 0
+	newlyUnassigned := len(newProjects) == 0
+
+	// because we don't do patchy updates, users will need to submit the projects
+	// field on update even if they are not changing it. While a user is often
+	// not allowed to _create_ an unassigned resource, they may be permitted to
+	// change the name of one, provided the projects are not altered. In those
+	// cases, we do not want to check authorization for moving resources to
+	// being unassigned.
+	if isUpdateRequest && previouslyUnassigned && newlyUnassigned {
```

Comment:
We could also just append `(unassigned)` into both arrays and continue with the work below (as the diff would end up empty), but seemed worth saving us a little work.